# **Group Assignment 03**

**Names**

**M. Anas Nasir (FA23-BBD-086)**

**M. Abdullah (SP24-BBD-089)**

**Ibtisam Ahmed (SP24-BBD-064)**

**M. Haseeb Waqas (FA23-BBD-095)**

**Usman Sabih (SP24-BBD-147)**

**Class BDA-4C**

----

In [1]:
import pandas as pd
import numpy as np

In [2]:
cars = pd.read_csv('cars.csv')

In [3]:
cars.head(15)

,Make,Model,Type,Origin,DriveTrain,MSRP,Invoice,EngineSize,Cylinders,Horsepower,MPG_City,MPG_Highway,Weight,Wheelbase,Length
0,Acura,MDX,SUV,Asia,All,36945,33337,3.5,6.0,265,17,23,4451,106,189
1,Acura,RSX Type S 2dr,Sedan,Asia,Front,23820,21761,2.0,4.0,200,24,31,2778,101,172
2,Acura,TSX 4dr,Sedan,Asia,Front,26990,24647,2.4,4.0,200,22,29,3230,105,183
3,Acura,TL 4dr,Sedan,Asia,Front,33195,30299,3.2,6.0,270,20,28,3575,108,186
4,Acura,3.5 RL 4dr,Sedan,Asia,Front,43755,39014,3.5,6.0,225,18,24,3880,115,197
5,Acura,3.5 RL w/Navigation 4dr,Sedan,Asia,Front,46100,41100,3.5,6.0,225,18,24,3893,115,197
6,Acura,NSX coupe 2dr manual S,Sports,Asia,Rear,89765,79978,3.2,6.0,290,17,24,3153,100,174
7,Audi,A4 1.8T 4dr,Sedan,Europe,Front,25940,23508,1.8,4.0,170,22,31,3252,104,179
8,Audi,A41.8T convertible 2dr,Sedan,Europe,Front,35940,32506,1.8,4.0,170,23,30,3638,105,180
9,Audi,A4 3.0 4dr,Sedan,Europe,Front,31840,28846,3.0,6.0,220,20,28,3462,104,179


In [4]:
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 428 entries, 0 to 427
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Make         428 non-null    object 
 1   Model        428 non-null    object 
 2   Type         428 non-null    object 
 3   Origin       428 non-null    object 
 4   DriveTrain   428 non-null    object 
 5   MSRP         428 non-null    int64  
 6   Invoice      428 non-null    int64  
 7   EngineSize   428 non-null    float64
 8   Cylinders    426 non-null    float64
 9   Horsepower   428 non-null    int64  
 10  MPG_City     428 non-null    int64  
 11  MPG_Highway  428 non-null    int64  
 12  Weight       428 non-null    int64  
 13  Wheelbase    428 non-null    int64  
 14  Length       428 non-null    int64  
dtypes: float64(2), int64(8), object(5)
memory usage: 50.3+ KB


In [5]:
cars.tail(15)

,Make,Model,Type,Origin,DriveTrain,MSRP,Invoice,EngineSize,Cylinders,Horsepower,MPG_City,MPG_Highway,Weight,Wheelbase,Length
413,Volkswagen,Jetta GL,Wagon,Europe,Front,19005,17427,2.0,4.0,115,24,30,3034,99,174
414,Volkswagen,Passat GLS 1.8T,Wagon,Europe,Front,24955,22801,1.8,4.0,170,22,31,3338,106,184
415,Volkswagen,Passat W8,Wagon,Europe,Front,40235,36956,4.0,8.0,270,18,25,4067,106,184
416,Volvo,XC90 T6,SUV,Europe,All,41250,38851,2.9,6.0,268,15,20,4638,113,189
417,Volvo,S40 4dr,Sedan,Europe,Front,25135,23701,1.9,4.0,170,22,29,2767,101,178
418,Volvo,S60 2.5 4dr,Sedan,Europe,All,31745,29916,2.5,5.0,208,20,27,3903,107,180
419,Volvo,S60 T5 4dr,Sedan,Europe,Front,34845,32902,2.3,5.0,247,20,28,3766,107,180
420,Volvo,S60 R 4dr,Sedan,Europe,All,37560,35382,2.5,5.0,300,18,25,3571,107,181
421,Volvo,S80 2.9 4dr,Sedan,Europe,Front,37730,35542,2.9,6.0,208,20,28,3576,110,190
422,Volvo,S80 2.5T 4dr,Sedan,Europe,All,37885,35688,2.5,5.0,194,20,27,3691,110,190


In [6]:
cars.shape

(428, 15)

There are 428 rows and 15 columns in the dataset.

The cars dataset is a mix of categorical and numerical variables, their classification is as follows.

<u>**Categorical Variables**</u>

1. **Make** is a categorical variable becuase it contains categories like Volkswagen, Volvo, Acura, Audi.
2. **Model** is also categorical variable and it contains different models of cars like Jetta GL, S80 2.54dr, MDX, RSX Type 2 sdr, A6 3.0 4dr etc.
3. **Type** is a categorical variable and it contains different segments of cars like Sedan, SUV, Wagon etc.
4. **Origin** tells us about the location where each car was manufactured like Europe, Asia, and USA.
5. **DriveTrain** tells us about the cars that are Rear Wheel Drive, Front Wheel Drive and All Wheel Drive.

<u>**Numerical Variables**</u>

1. **MSRP** stands for Manufactured Suggested Retail Price (Price recommended by the manufacturer) , and is different for each make and model of cars.
2. **Invoice** is the amount the dealer pays to the manufacturer and like **MSRP** it is different for each make and model of cars.
3. **EngineSize** tells us about the size of engine in Liters of CC, in our dataset it is, 3.5, 2.0, 2.4 etc.
4. **Cylinders** tells us about the presence of cylinders in our cars dataset, for example, **Acura MDX** has **6 cylinders** present in it, similarly **Acura TL 4dr** also has 6 cylinders.
5. **Horsepower** measures the output of engine size, 265, 200, 270 etc.
6. **MPG (Highway and City)** Mileage of every car when it is in driven in city conditions and in highway conditions.
7. **Weight** represents the weight of each car as the name suggests.
8. **Wheelbase** is the distance between the centres of the front and rear wheels of a vehicle.
9. **Length** is the total exterior length of a vehicle i.e., the distance from front to rear bumper, usually measured in inches or millimeters

-----

## <u>**Parametric and Non-parametric tests**</u>


## <u>**Parametric Test**</u>

Parametric test is a statistical technique that makes assumptions on basis of how much your data distribution is normal. It is suitable for quantitiatve (numeric) and continuous variables. They are numeric quantitities that represent whole population. 


## <u>**Non-Parametric Test**</u>

The test that doesn't rely on any statistical distribution; and hence the name suggets it is also known as "distribution-free" hypothesis test. These tests do not have parameters of the population. Non-parametric test is suitable for variables that are categorical in nature.

## **<u>Introduction to Hypothesis</u>**

A hypothesis is an assumption about a population that we test using statistical methods. It helps us decide whether the observed data supports or goes against them.

There are two types of hypothesis to claim whether our test supports or goes against it.

<u>**Null Hypothesis (Ho)**</u>

The hypothesis test that assumes that there is no effect or relation, or no difference. 

<u>**Alternate Hypothesis (Ha)**</u>

The hypothesis that assumes that a relationship, effect or difference exists.

<u>**Types of Parametic Tests**</u>

1. **t-test:** A specific type of test that is used for checking if there is a significant difference between means of the two groups concerned. It is categorized into two types:

     **i. One-sample t-test** is used to check if there is a significant difference a sample and hypothesized population means.

     **ii. Two-sample t-test** is used to measure significant difference between two groups.   


2. **ANOVA (Analysis of Variance)**

     ANOVA is a statistical inference test used for comparing multiple groups. It usually compares more than two datasets. It has variations like **One-way ANOVA** (Multile group comparison using single independent variable), **Two-way ANOVA** (multiple groups comparison based on two independet variables), and the last **N-way Anova** (Multiple groups comparison based on N independent variables).

**<u>Type of Non-Parametric Test</u>**

**Chi-square test**, a test that is determined by a significant difference or relationship between two categorical variables from single population. In general, this test assesses whether distribution of categorical variables differ from each other. It is also known as Chi-square goodness of fit test or test for independence. 

## **<u> Statistical parametric tests</u>**

## <u>**One-sample t-test**</u>

In [1]:
import pandas as pd
from scipy import stats

In [2]:
cars = pd.read_csv('cars.csv')

### **Targeted variables of One-sample t-test**

1. DriveTrain 
2. MSRP

**Hypothesis**: To examine whether the mean MSRP differs significantly between Front-wheel and Rear-wheel drive vehicles.

In [13]:
print('\nTest 1: T-test on MSRP by DriveTrain (Front vs Rear)')

front = cars[cars['DriveTrain'] == 'Front']['MSRP']
rear = cars[cars['DriveTrain'] == 'Rear']['MSRP']
t_stat1, p_val1 = stats.ttest_ind(front, rear, equal_var = False)

print(f"T-statistic = {t_stat1:.3f}")
print(f"P-value = {p_val1: .5f}")


Test 1: T-test on MSRP by DriveTrain (Front vs Rear)
T-statistic = -7.896
P-value =  0.00000


## **Decision**

In [15]:
if p_val1 < 0.05:
    print("Reject the null hypothesis: There is a significant difference in mean MSRP between Front-wheel and Rear-wheel drive vehicles, alternate hypothesis will be accepted.")
else:
    print("Fail to reject the null hypothesis: No significant difference in mean MSRP between Front-wheel and Rear-wheel drive vehicles, null hypothesis will be accepted.")

Reject the null hypothesis: There is a significant difference in mean MSRP between Front-wheel and Rear-wheel drive vehicles, alternate hypothesis will be accepted.


## **<u>Two Sample T-test**</u>

## **Variables targeted**

1. MSRP
2. Origin

## **Hypothesis**

To evaluate whether cars manufactured in Asia and USA differ in their mean MSRP

In [17]:
print('\n Test 2: T-test on MSRP by Origin (Asia vs USA)')

asia = cars[cars['Origin'] == 'Asia']['MSRP']
usa = cars[cars['Origin'] == 'USA']['MSRP']

t_stat2, p_val2 = stats.ttest_ind(asia, usa, equal_var = False)

print(f"T-statistic = {t_stat2:.3f}")
print(f"P-value = {p_val2: .5f}")

if p_val2 < 0.05:
    print("Reject the null hypothesis: There is a significant difference in mean MSRP between cars manufactured in Asia and USA, alternate hypothesis will be accepted.")
else:
     print("Fail to reject the null hypothesis: No significant difference in mean MSRP between cars manufactured in Asia and USA, null hypothesis will be accepted.")



 Test 2: T-test on MSRP by Origin (Asia vs USA)
T-statistic = -2.753
P-value =  0.00626
Reject the null hypothesis: There is a significant difference in mean MSRP between cars manufactured in Asia and USA, alternate hypothesis will be accepted.


## <u>**ANOVA (Analysis of Variance)**</u>

## **Hypothesis**

To determine whether the mean **MSRP** differs significantly among cars from different **Origin** (Asia, Europe, USA).

## <u>**One-way ANOVA**</u>


In [19]:
cars.dropna(subset = ['Origin', 'MSRP'])

asia = cars[cars['Origin'] == 'Asia']['MSRP']
usa = cars[cars['Origin'] == 'USA']['MSRP']
europe = cars[cars['Origin'] == 'Europe']['MSRP']

f_stat, p_val = stats.f_oneway(asia, usa, europe)

print(f"F-statistic = {f_stat:.3f}")
print(f"P-value = {p_val: .5f}")

F-statistic = 76.983
P-value =  0.00000


In [20]:
if p_val1 < 0.05:
    print("Reject the null hypothesis: There is a significant difference in mean MSRP between Front-wheel and Rear-wheel drive vehicles.")
else:
     print("Fail to reject the null hypothesis: No significant difference in mean MSRP between Front-wheel and Rear-wheel drive vehicles.")

Reject the null hypothesis: There is a significant difference in mean MSRP between Front-wheel and Rear-wheel drive vehicles.


## <u>**Two-way ANOVA**</u>

## **Hypothesis**

To determine whether **Origin** and DriveTrain individually and collectively influence the **MSRP** of cars.

In [23]:
import statsmodels.api as sm
from statsmodels.formula.api import ols

cars = cars.dropna(subset = ['Origin', 'MSRP', 'DriveTrain'])

model = ols('MSRP ~ C(Origin) + C(DriveTrain) + C(Origin):C(DriveTrain)', data = cars).fit()
anova_table = sm.stats.anova_lm(model, typ = 2)

print(anova_table)

                               sum_sq     df          F        PR(>F)
C(Origin)                2.550375e+10    2.0  55.373753  4.585630e-22
C(DriveTrain)            1.784142e+10    2.0  38.737312  3.659355e-16
C(Origin):C(DriveTrain)  4.023092e+09    4.0   4.367471  1.802457e-03
Residual                 9.649039e+10  419.0        NaN           NaN


In [24]:
print('\nHo1: No difference in mean MSRP between cars manufactured in Asia, USA, and Europe.')
print('Ho2: No difference in mean MSRP between Front-wheel and Rear-wheel drive vehicles.')
print('Ho3: No interaction effect between Origin and DriveTrain on mean MSRP.')

print('\nDecision:')

for factor, p in anova_table['PR(>F)'].items():
    if p < 0.05:
        print(f"Reject the null hypothesis for {factor}: Significant effect on mean MSRP.")
    else:
        print(f"Fail to reject the null hypothesis for {factor}: No significant effect on mean MSRP.")


Ho1: No difference in mean MSRP between cars manufactured in Asia, USA, and Europe.
Ho2: No difference in mean MSRP between Front-wheel and Rear-wheel drive vehicles.
Ho3: No interaction effect between Origin and DriveTrain on mean MSRP.

Decision:
Reject the null hypothesis for C(Origin): Significant effect on mean MSRP.
Reject the null hypothesis for C(DriveTrain): Significant effect on mean MSRP.
Reject the null hypothesis for C(Origin):C(DriveTrain): Significant effect on mean MSRP.
Fail to reject the null hypothesis for Residual: No significant effect on mean MSRP.


## <U>**N-Way ANOVA**</U>

## **Hypothesis**

To investigate how **Origin**, **DriveTrain**, and **Cylinders** collectively and interactibely affect car prices.

In [26]:
cars = cars.dropna(subset = ['Origin', 'MSRP', 'DriveTrain', 'Cylinders'])

model = ols('MSRP ~ C(Origin) + C(DriveTrain) + C(Cylinders) + C(Origin):C(DriveTrain) + C(Origin):C(Cylinders) + C(DriveTrain):C(Cylinders) + C(Origin):C(DriveTrain):C(Cylinders)', data = cars).fit()
anova_table = sm.stats.anova_lm(model, typ = 2)

print(anova_table)

print('\nHypotheses:')

print('Ho1: None of the main or interaction effects significantly affect mean MSRP.')
print('Ho2: At least one main or interaction effect significantly affects mean MSRP.')

print('\nDecision:')

for factor, p in anova_table['PR(>F)'].items():
     if p < 0.05:
          print(f"Reject the null hypothesis for {factor}: Significant effect on mean MSRP.")
     else:
          print(f"Fail to reject the null hypothesis for {factor}: No significant effect on mean MSRP.")


                                            sum_sq     df             F  \
C(Origin)                             6.144261e-03    2.0  2.259238e-11   
C(DriveTrain)                         0.000000e+00    2.0  0.000000e+00   
C(Cylinders)                         -3.603496e+10    6.0 -4.416671e+01   
C(Origin):C(DriveTrain)               4.676530e+08    4.0  8.597774e-01   
C(Origin):C(Cylinders)                4.624331e+10   12.0  2.833935e+01   
C(DriveTrain):C(Cylinders)            1.020579e+10   12.0  6.254430e+00   
C(Origin):C(DriveTrain):C(Cylinders)  2.463837e+10   24.0  7.549585e+00   
Residual                              5.330449e+10  392.0           NaN   

                                            PR(>F)  
C(Origin)                             9.999962e-01  
C(DriveTrain)                         1.000000e+00  
C(Cylinders)                          1.000000e+00  
C(Origin):C(DriveTrain)               4.240529e-01  
C(Origin):C(Cylinders)                3.909309e-28  
C(Driv

c:\Users\manas\AppData\Local\Programs\Python\Python313\Lib\site-packages\statsmodels\base\model.py:1894: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 2, but rank is 1
  warnings.warn('covariance of constraints does not have full '
c:\Users\manas\AppData\Local\Programs\Python\Python313\Lib\site-packages\statsmodels\base\model.py:1894: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 2, but rank is 1
  warnings.warn('covariance of constraints does not have full '
c:\Users\manas\AppData\Local\Programs\Python\Python313\Lib\site-packages\statsmodels\base\model.py:1894: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 6, but rank is 2
  warnings.warn('covariance of constraints does not have full '
c:\Users\manas\AppData\Local\Programs\Python\Python313\Lib\site-packages\statsmodels\base\model.py:1894: ValueWarning: covariance of constraints does not have f

## <u>**Non-Parametric Test**</u>

## **Chi-Sqare Test**

## **Hypothesis**

To determine if there is an association between the type of DriveTrain (Front, Rear, All) and the Origin of the car (Asia, USA, Europe).

In [30]:
from scipy.stats import chi2_contingency

cars = cars.dropna(subset = ['Origin', 'DriveTrain'])

contingency_table = pd.crosstab(cars['Origin'], cars['DriveTrain'])

print("Contingency Table")

chi2, p_val, dof, expected = chi2_contingency(contingency_table)

print(f"Chi-square statistic = {chi2:.3f}")
print(f"P-value = {p_val:.5f}")
print(f'Degrees of Freedom = {dof}')
print(F"P-value = {p_val:.5f}")

pd.DataFrame(expected, index = contingency_table.index, columns = contingency_table.columns)

Contingency Table
Chi-square statistic = 42.036
P-value = 0.00000
Degrees of Freedom = 4
P-value = 0.00000


DriveTrain,All,Front,Rear
Origin,,,
Asia,33.690141,82.760563,39.549296
Europe,26.563380,65.253521,31.183099
USA,31.746479,77.985915,37.267606


In [31]:
print('\nHo: Origin and DriveTrain are independent.')
print('Ha: Origin and DriveTrain are not independent.')

print('\nDecision:')

if p_val < 0.05:
     print("Reject the null hypothesis: There is a significant association between Origin and DriveTrain.")
else:
     print("Fail to reject the null hypothesis: No significant association between Origin and DriveTrain.")
     


Ho: Origin and DriveTrain are independent.
Ha: Origin and DriveTrain are not independent.

Decision:
Reject the null hypothesis: There is a significant association between Origin and DriveTrain.


-----------